src/libraries/Common/src/System/Net/MultiArrayBuffer.cs

```diff
@@ -0,0 +1,431 @@
+                Debug.Assert(blocks is not null);
+                Debug.Assert(start <= int.MaxValue);
+                Debug.Assert(length <= int.MaxValue);
+                Debug.Assert(start + length <= blocks.Length * BlockSize);
+
+                _blocks = blocks;
+                _start = start;
+                _length = length;
+            }
+        }
+
+        private static uint GetBlockIndex(uint offset) => offset / BlockSize;
+        private static uint GetOffsetInBlock(uint offset) => offset % BlockSize;
+
+        public bool IsEmpty => _length == 0;
+
+        public int Length => (int)_length;
+
+        public ref byte this[int index]
+        {
+            get
+            {
+                uint uindex = (uint)index;
+                if (uindex >= _length)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+
+                uint offset = _start + uindex;
+                return ref _blocks![GetBlockIndex(offset)]![GetOffsetInBlock(offset)];
+            }
+        }
+
+        public int BlockCount => (int)(GetBlockIndex(_start + _length + (BlockSize - 1)) - GetBlockIndex(_start));
+
+        public Memory<byte> GetBlock(int blockIndex)
+        {
+            if ((uint)blockIndex >= BlockCount)
+            {
+                throw new IndexOutOfRangeException();
+            }
+
+            Debug.Assert(_length > 0, "Length should never be 0 here because BlockCount would be 0");
+            Debug.Assert(_blocks is not null);
+
+            uint startInBlock = (blockIndex == 0 ? GetOffsetInBlock(_start) : 0);
+            uint endInBlock = (blockIndex == BlockCount - 1 ? GetOffsetInBlock(_start + _length - 1) + 1 : BlockSize);
+
+            Debug.Assert(0 <= startInBlock, $"Invalid startInBlock={startInBlock}. blockIndex={blockIndex}, _blocks.Length={_blocks.Length}, _start={_start}, _length={_length}");
+            Debug.Assert(startInBlock < endInBlock, $"Invalid startInBlock={startInBlock}, endInBlock={endInBlock}. blockIndex={blockIndex}, _blocks.Length={_blocks.Length}, _start={_start}, _length={_length}");
+            Debug.Assert(endInBlock <= BlockSize, $"Invalid endInBlock={endInBlock}. blockIndex={blockIndex}, _blocks.Length={_blocks.Length}, _start={_start}, _length={_length}");
+
+            return new Memory<byte>(_blocks[GetBlockIndex(_start) + blockIndex], (int)startInBlock, (int)(endInBlock - startInBlock));
+        }
+
+        public MultiMemory Slice(int start)
+        {
+            uint ustart = (uint)start;
+            if (ustart > _length)
+            {
+                throw new IndexOutOfRangeException();
+            }
+
+            return new MultiMemory(_blocks, _start + ustart, _length - ustart);
+        }
+
+        public MultiMemory Slice(int start, int length)
+        {
+            uint ustart = (uint)start;
+            uint ulength = (uint)length;
+            if (ustart > _length || ustart + ulength > _length)
```

Comment:
Is the first check needed?

Reply:
No, good catch

Reply:
Ah, actually, I remember why I did this.

We aren't explicitly checking here if `start >= 0`, but if it is, this will get caught in the `ustart > _length` test.

Unfortunately the second check here isn't quite right, since it doesn't catch a case like start = 5, length = -1, I think it should instead be `ulength > _length - ustart` and that should correctly catch all invalid cases.

This is also why we don't have asserts here, as you asked about above.

Note also there are asserts in the MultiMemory constructor, so if we miss something here it will get asserted there.